alphacsc/_encoder.py

```diff
@@ -0,0 +1,336 @@
+import numpy as np
+from .loss_and_gradient import compute_X_and_objective_multi
+from .update_z_multi import update_z_multi
+from .utils import check_dimension, lil
+
+
+# XXX check consistency / proper use!
+def get_z_encoder_for(
+        solver,
+        z_kwargs,
+        X,
+        D_hat,
+        n_atoms,
+        atom_support,
+        algorithm,
+        reg,
+        loss,
+        loss_params,
+        uv_constraint,
+        feasible_evaluation,
+        n_jobs,
+        use_sparse_z):
+    """
+    Returns a z encoder for the required solver.
+
+    Parameters
+    ----------
+    solver : str
+        The solver to use for the z update. Options are
+        {{'l_bfgs' | 'lgcd'}}.
+    z_kwargs : dict
+        Additional keyword arguments to pass to update_z_multi.
+    X : array, shape (n_trials, n_channels, n_times)
+        The data on which to perform CSC.
+    D_hat : array, shape e(n_trials, n_channels, n_times) or
+        (n_atoms, n_channels + atom_support)
+        The dictionary used to encode the signal X. Can be either in the form
+        f a full rank dictionary D (n_atoms, n_channels, atom_support) or with
+        the spatial and temporal atoms uv (n_atoms, n_channels + atom_support)
+    n_atoms : int
+        The number of atoms to learn.
+    atom_support : int
+        The support of the atom.
+    algorithm : 'batch' | 'greedy' | 'online' | 'stochastic'
+        Dictionary learning algorithm.
+    reg : float
+        The regularization parameter.
+    loss : {{ 'l2' | 'dtw' | 'whitening'}}
+        Loss for the data-fit term. Either the norm l2 or the soft-DTW.
+    loss_params : dict
+        Parameters of the loss.
+    uv_constraint : {{'joint' | 'separate'}}
+        The kind of norm constraint on the atoms:
+
+        - :code:`'joint'`: the constraint is ||[u, v]||_2 <= 1
+        - :code:`'separate'`: the constraint is ||u||_2 <= 1 and ||v||_2 <= 1
+    feasible_evaluation : boolean
+        If feasible_evaluation is True, it first projects on the feasible set,
+        i.e. norm(uv_hat) <= 1.
+    n_jobs : int
+        The number of parallel jobs.
+    use_sparse_z : boolean
+        Use sparse lil_matrices to store the activations.
+
+    Returns
+    -------
+    enc - a ZEncoder instance
+
```

Comment:
```suggestion
    enc : instance of ZEncoder
        The encoder.

```

we should run numpydoc checks in the repo.

Reply:
OK, I'll add an issue for that